pkgs/stdenv/darwin/default.nix

```diff
@@ -314,495 +368,1016 @@ rec {
+              buildCommand = ''
+                mkdir -p $out/lib
+                ln -s ${bootstrapTools}/lib/libc++abi.dylib $out/lib
+              '';
+              passthru = {
+                libName = "c++abi";
+                isFromBootstrapFiles = true;
+              };
+            };
+            compiler-rt = self.stdenv.mkDerivation {
+              name = "bootstrap-stage0-compiler-rt";
+              buildCommand = ''
+                mkdir -p $out/lib $out/share
+                ln -s ${bootstrapTools}/lib/libclang_rt* $out/lib
+                ln -s ${bootstrapTools}/lib/darwin       $out/lib
+              '';
+              passthru.isFromBootstrapFiles = true;
+            };
+          });
+        in
+        # Define `release_version` as a `*`, so it’s interpreted as a glob. This is necessary
+        # because the version of clang in the bootstrap tools is not known ahead of time. This
+        # should be safe because the bootstrap tools should never contain multiple clangs.
+        { inherit tools libraries; release_version = "*"; } // tools // libraries
+      );
+    };
 
-        libcxx = stdenv.mkDerivation {
-          name = "bootstrap-stage0-libcxx";
-          dontUnpack = true;
-          installPhase = ''
-            mkdir -p $out/lib $out/include
-            ln -s ${bootstrapTools}/lib/libc++.dylib $out/lib/libc++.dylib
-            ln -s ${bootstrapTools}/include/c++      $out/include/c++
-          '';
-          passthru = {
-            isLLVM = true;
-            cxxabi = self."${finalLlvmPackages}".libcxxabi;
-          };
+    # The bootstrap tools may use `strip` from cctools, so use a compatible set of flags until LLVM
+    # is rebuilt, and darwin.binutils can use its implementation instead.
+    extraPreHook = ''
+      stripAllFlags=" "    # the cctools "strip" command doesn't know "-s"
+      stripDebugFlags="-S" # the cctools "strip" command does something odd with "-p"
+    '';
+  })
+
+  # This stage is primarily responsible for building the linker and setting up versions of
+  # certain dependencies needed by the rest of the build process. It is necessary to rebuild the
+  # linker because the `compiler-rt` build process checks the version and attempts to manually
+  # run `codesign` if it detects a version of `ld64` it considers too old. If that happens, the
+  # build process will fail for a few different reasons:
+  #  - sigtool is too old and does not accept the `--sign` argument;
+  #  - sigtool is new enough to accept the `--sign` argument, but it aborts when it is invoked on a
+  #    binary that is already signed; or
+  #  - compiler-rt attempts to invoke `codesign` on x86_64-darwin, but `sigtool` is not currently
+  #    part of the x86_64-darwin bootstrap tools.
+  #
+  # This stage also builds CF and Libsystem to simplify assertions and assumptions for later by
+  # making sure both packages are present on x86_64-darwin and aarch64-darwin.
```

Comment:
I appreciate the context from this detailed comment, which is the only reason I can ask these follow up questions:

1. Does this mean we can remove a rebuild of the linker if bootstrap tools can be updated one day to use a recent enough linker (or is the recency check a moving target so we should always keep this here)?

2. It's not obvious to me how CF and Libsystem are being built in this stage (I guess it's because I don't see any obvious references to either in the code).

Reply:
One of the goals of the rework is to be decoupled from what’s in the bootstrap, so I’d say no. Even if the bootstrap tools are updated with the latest versions, it seems like it would be better to stay decoupled to allow future updates to be made independently of what’s in the bootstrap tools.

Regarding rebuilds, if something is not specified in the overlay, it will be rebuilt (as needed due to lazy evaluation).